### PR TITLE
fix: allow using build env in kustomize images

### PR DESCRIPTION
**Pull request by globin**
This allows substituting the target revision using image:${ARGOCD_APP_REVISION}

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [x] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [x] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)). 
* [x] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.

Please see [Contribution FAQs](https://argo-cd.readthedocs.io/en/latest/developer-guide/faq/) if you have questions about your pull-request.

### CODE REVIEW

File: util/kustomize/kustomize.go [hunk 1]
```diff
@@ -108,7 +108,9 @@ func (k *kustomize) Build(opts *v1alpha1.ApplicationSourceKustomize, kustomizeOp
 			// set image node:8.15.0 mysql=mariadb alpine@sha256:24a0c4b4a4c0eb97a1aabb8e29f18e917d05abfe1b7a7c07857230879ce7d3d3
 			args := []string{"edit", "set", "image"}
 			for _, image := range opts.Images {
-				args = append(args, string(image))
+				// this allows using ${ARGOCD_APP_REVISION}
+				envSubstitutedImage := envVars.Envsubst(string(image))
```
Comment (crenshaw-dev) [inline util/kustomize/kustomize.go:112]: Do we need to be worried about dollar signs in existing image field values? Not sure what's possible in that field.

e.g.

```yaml
- images:
  service=big$money-no-whammies  # I mean this literally, please don't evaluate to `big-no-whammies`.
```

Comment (globin) [reply util/kustomize/kustomize.go:112]: Quoting from the docker documentation:
> Name components may contain lowercase letters, digits and separators. A separator is defined as a period, one or two underscores, or one or more dashes. A name component may not start or end with a separator.
> 
> A tag name must be valid ASCII and may contain lowercase and uppercase letters, digits, underscores, periods and dashes. A tag name may not start with a period or a dash and may contain a maximum of 128 characters.

https://docs.docker.com/engine/reference/commandline/tag/#extended-description

Therefore this case would never happen with a well-formed image name & tag.

**Review by crenshaw-dev — APPROVED** [review]

**Comment by globin** [comment]
2.7 would be nice if possible, I'm currently using a custom-patched image.

**Comment by crenshaw-dev** [comment]
It'll be in 2.7.4. Thanks for the PR!

**Comment by globin** [comment]
Thank you very much!

**Pull request merged**